File: solution/143. Reorder List/143_02.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode* head) {
        if (!head) return;
        
        // 1. find the middle node
        auto *slow = head, *fast = head;
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        
        // 2. cut and reverse the second list
        ListNode *head2 = nullptr;
        ListNode *last = slow->next;
        slow->next = nullptr;
        while (last) {
            auto *tmp = last->next;
            last->next = head2;
            head2 = last;
            last = tmp;
        }
        
        // 3. merge two list
        while (head && head2) {
            auto *tmp1 = head->next, *tmp2 = head2->next;
            head->next = head2;
            head2->next = tmp1;
            head = tmp1;
            head2 = tmp2;
        }
    }
};

File: solution/143. Reorder List/143_02.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public void reorderList(ListNode head) {
        if (head == null) return;
        
        ListNode slow = head, fast = head;
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        ListNode p2 = reverse(slow.next);
        slow.next = null;
        
        ListNode p1 = head;
        while (p1 != null && p2 != null) {
            ListNode tmp1 = p1.next;
            ListNode tmp2 = p2.next;
            p1.next = p2;
            p2.next = tmp1;
            p1 = tmp1;
            p2 = tmp2;
        }
    }
    
    private ListNode reverse(ListNode node) {
        ListNode pre = null;
        while (node != null) {
            ListNode tmp = node.next;
            node.next = pre;
            pre = node;
            node = tmp;
        }
        return pre;
    }
}
